### PR TITLE
Test dumper change to include Isimplicit/Language and some formatting change 

### PR DESCRIPTION
**Pull request by jinujoseph**
fixes https://github.com/dotnet/roslyn/issues/21330

### CODE REVIEW

**Comment by jinujoseph** [comment]
test windows_debug_vs-integration_prtest

**Comment by jinujoseph** [comment]
test windows_release_vs-integration_prtest

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests.cs [hunk 1]
```diff
@@ -46,28 +46,32 @@ static void Test2(int y, params int[] x)
             var nodes = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().ToArray();
 
             compilation.VerifyOperationTree(nodes[0], expectedOperationTree:
-@"IInvocationExpression (void Cls.Test1(params System.Int32[] x)) (OperationKind.InvocationExpression, Type: System.Void) (Syntax: 'Test1(null)')
-  Instance Receiver: null
+@"IInvocationExpression (void Cls.Test1(params System.Int32[] x)) (OperationKind.InvocationExpression, Type: System.Void, Language: C#) (Syntax: 'Test1(null)')
+  Instance Receiver: 
+  null
```
Comment (mavasani) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests.cs]: This needs an `Indent`

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs [hunk 2]
```diff
@@ -22,10 +22,11 @@ public void F()
 }
 ";
             string expectedOperationTree = @"
-IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: System.String[]) (Syntax: 'new string[1]')
+IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: System.String[], Language: C#) (Syntax: 'new string[1]')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: >Language: C# [](start = 88, length = 12)

It is probably not that useful to dump Language for all nodes in the tree. We could pass the expected language to the VerifyOperationTreeAndDiagnosticsForTest helper and have it verify that all nodes use it.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: Perhaps VerifyOperationTreeAndDiagnosticsForTest could even figure out what language to expect based on the syntax node, or, if it is already language specific, we can hardcode expected language in it.

---
In reply to: [141148813](https://github.com/dotnet/roslyn/pull/22309#discussion_r141148813) [](ancestors = 141148813)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs [hunk 3]
```diff
@@ -504,37 +531,44 @@ public void F()
...
                         Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
-                        Operand: ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 3) (Syntax: '3')
+                        Operand: 
+                          ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 3) (Syntax: '3')
           IArrayInitializer (1 elements) (OperationKind.ArrayInitializer) (Syntax: '{ { 4, 5, 6 } }')
             Element Values(1):
                 IArrayInitializer (3 elements) (OperationKind.ArrayInitializer) (Syntax: '{ 4, 5, 6 }')
                   Element Values(3):
-                      IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Byte, Constant: 4) (Syntax: '4')
+                      IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Byte, Constant: 4, IsImplicit) (Syntax: '4')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_ArrayCreationAndInitializer.cs]: >Implicit [](start = 45, length = 8)

Perhaps the test visitor should assert that Implicit == IsImplicit and baseline should output just one property. Should also consider simply removing the Implicit property.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs [hunk 4]
```diff
@@ -995,17 +1034,18 @@ void M1()
 ";
             string expectedOperationTree = @"
 IInvocationExpression ( void P.M2([System.String memberName = null], [System.String sourceFilePath = null], [System.Int32 sourceLineNumber = 0])) (OperationKind.InvocationExpression, Type: System.Void) (Syntax: 'M2()')
-  Instance Receiver: IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: P) (Syntax: 'M2')
+  Instance Receiver: 
+    IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: P, IsImplicit) (Syntax: 'M2')
   Arguments(3):
-      IArgument (ArgumentKind.DefaultValue, Matching Parameter: memberName) (OperationKind.Argument) (Syntax: 'M2()')
+      IArgument (ArgumentKind.DefaultValue, Matching Parameter: memberName) (OperationKind.Argument, IsImplicit) (Syntax: 'M2()')
         ILiteralExpression (OperationKind.LiteralExpression, Type: System.String, Constant: ""M1"") (Syntax: 'M2()')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs]: >ILiteralExpression (OperationKind.LiteralExpression, Type: System.String, Constant: ""M1"") (Syntax: 'M2()') [](start = 8, length = 108)

I think values for arguments with ArgumentKind.DefaultValue should be marked as Implicit.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs [hunk 5]
```diff
@@ -652,17 +672,19 @@ void M1()
...
+  Instance Receiver: 
+    IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: P, IsImplicit) (Syntax: 'M2')
   Arguments(2):
       IArgument (ArgumentKind.Explicit, Matching Parameter: x) (OperationKind.Argument) (Syntax: '1')
         ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 1) (Syntax: '1')
         InConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
         OutConversion: CommonConversion (Exists: True, IsIdentity: True, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
-      IArgument (ArgumentKind.ParamArray, Matching Parameter: array) (OperationKind.Argument) (Syntax: 'M2(1, 0.1, 0.2)')
+      IArgument (ArgumentKind.ParamArray, Matching Parameter: array) (OperationKind.Argument, IsImplicit) (Syntax: 'M2(1, 0.1, 0.2)')
         IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: System.Double[]) (Syntax: 'M2(1, 0.1, 0.2)')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs]: >IArrayCreationExpression (OperationKind.ArrayCreationExpression, Type: System.Double[]) (Syntax: 'M2(1, 0.1, 0.2)') [](start = 8, length = 115)

I think this node should be marked with IsImplicit.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs [hunk 6]
```diff
@@ -652,17 +672,19 @@ void M1()
...
           Dimension Sizes(1):
               ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 2) (Syntax: 'M2(1, 0.1, 0.2)')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs]: >ILiteralExpression (OperationKind.LiteralExpression, Type: System.Int32, Constant: 2) (Syntax: 'M2(1, 0.1, 0.2)') [](start = 14, length = 113)

I think this node should be marked with IsImplicit.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs [hunk 7]
```diff
@@ -652,17 +672,19 @@ void M1()
...
-          Initializer: IArrayInitializer (2 elements) (OperationKind.ArrayInitializer) (Syntax: 'M2(1, 0.1, 0.2)')
+          Initializer: 
+            IArrayInitializer (2 elements) (OperationKind.ArrayInitializer) (Syntax: 'M2(1, 0.1, 0.2)')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IArgument.cs]: >IArrayInitializer (2 elements) (OperationKind.ArrayInitializer) (Syntax: 'M2(1, 0.1, 0.2)') [](start = 12, length = 91)

I think this node should be marked with IsImplicit.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IEventAssignmentExpression.cs [hunk 8]
```diff
@@ -36,10 +36,14 @@ void M()
...
-  Event Reference: IEventReferenceExpression: event System.EventHandler Test.MyEvent (OperationKind.EventReferenceExpression, Type: System.EventHandler) (Syntax: 't.MyEvent')
-      Instance Receiver: ILocalReferenceExpression: t (OperationKind.LocalReferenceExpression, Type: Test) (Syntax: 't')
-  Handler: IMethodReferenceExpression: void C.Handler(System.Object sender, System.EventArgs e) (OperationKind.MethodReferenceExpression, Type: System.EventHandler) (Syntax: 'Handler')
-      Instance Receiver: IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: C) (Syntax: 'Handler')
+  Event Reference: 
+    IEventReferenceExpression: event System.EventHandler Test.MyEvent (OperationKind.EventReferenceExpression, Type: System.EventHandler) (Syntax: 't.MyEvent')
+      Instance Receiver: 
+        ILocalReferenceExpression: t (OperationKind.LocalReferenceExpression, Type: Test) (Syntax: 't')
+  Handler: 
+    IMethodReferenceExpression: void C.Handler(System.Object sender, System.EventArgs e) (OperationKind.MethodReferenceExpression, Type: System.EventHandler) (Syntax: 'Handler')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IEventAssignmentExpression.cs]: >IMethodReferenceExpression: void C.Handler(System.Object sender, System.EventArgs e) (OperationKind.MethodReferenceExpression, Type: System.EventHandler) (Syntax: 'Handler') [](start = 4, length = 173)

It feels like the handler should be an Implicit delegate creation expression based on IMethodReferenceExpression and the type of the IMethodReferenceExpression should be null.

**Comment by AlekseyTs** [comment]
>     public partial class IOperationTests : SemanticModelTestBase

I think we should also test scenarios when right hand side of the assignment is a delegate instance.

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IEventAssignmentExpression.cs:10 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                 Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "12.0").WithArguments("double", "int?").WithLocation(8, 28)

@gafter It looks like this error is somewhat misleading. Should it say "Cannot implicitly convert type 'double' to **'int'**"? Instead of 'int?'. At least the tree seems to indicate that we are converting to the underlying type.

---
Refers to: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IIsPatternExpression.cs:239 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IWhileUntilLoopStatement.cs [hunk 9]
```diff
@@ -1035,24 +1257,31 @@ static bool TakeOutParam<T>(T y, out T x)
...
               Conversion: CommonConversion (Exists: True, IsIdentity: False, IsNumeric: False, IsReference: False, IsUserDefined: False) (MethodSymbol: null)
-              Operand: IInvocationExpression (System.Boolean X.TakeOutParam<System.Int32>(System.Int32 y, out System.Int32 x)) (OperationKind.InvocationExpression, Type: System.Boolean) (Syntax: 'TakeOutPara ... out var x1)')
-                  Instance Receiver: null
+              Operand: 
+                IInvocationExpression (System.Boolean X.TakeOutParam<System.Int32>(System.Int32 y, out System.Int32 x)) (OperationKind.InvocationExpression, Type: System.Boolean) (Syntax: 'TakeOutPara ... out var x1)')
+                  Instance Receiver: 
+                    null
                   Arguments(2):
-                      IArgument (ArgumentKind.Explicit, Matching Parameter: y) (OperationKind.Argument) (Syntax: 'f ? 1 : 2')
+                      IArgument (ArgumentKind.Explicit, Matching Parameter: y) (OperationKind.Argument, IsImplicit) (Syntax: 'f ? 1 : 2')
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IWhileUntilLoopStatement.cs]: >IArgument (ArgumentKind.Explicit, Matching Parameter: y) (OperationKind.Argument, IsImplicit) (Syntax: 'f ? 1 : 2') [](start = 22, length = 115)

This looks incorrect, the argument is explicit and it should be associated with an ArgumentSyntax.

**Comment by AlekseyTs** [comment]
>           IInvocationExpression ( ? MyList<System.String>.Add()) (OperationKind.InvocationExpression, Type: ?, IsInvalid, IsImplicit) (Syntax: '""str""')

I would expect to get ICollectionElementInitializerExpression node instead, even for an error case.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs:791 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>           IInvalidExpression (OperationKind.InvalidExpression, Type: ?, IsInvalid, IsImplicit) (Syntax: '""""')

Should this be a ICollectionElementInitializerExpression node?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs:844 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                   IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: MemberInitializerTest) (Syntax: 'z')

It feels like this node should be marked as implicit.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs:1077 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                   IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: Test, IsInvalid) (Syntax: 'Prop')

Should this node be implicit?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs:1155 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                             IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: MemberInitializerTest) (Syntax: 'x')

Should this node be implicit?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs:1163 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                             IInstanceReferenceExpression (OperationKind.InstanceReferenceExpression, Type: MemberInitializerTest) (Syntax: 'y')

Should this node be implicit?

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/ObjectAndCollectionInitializerTests.cs:1170 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                         IOperation:  (OperationKind.None) (Syntax: 'i')

OperationKind.None is unexpected, especially for success scenario.

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/QueryTests.cs:75 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                         IOperation:  (OperationKind.None) (Syntax: 'i')

This issue affects many tests in this file.

---
In reply to: [333984680](https://github.com/dotnet/roslyn/pull/22309#issuecomment-333984680) [](ancestors = 333984680)

---
Refers to: src/Compilers/CSharp/Test/Semantic/Semantics/QueryTests.cs:75 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

File: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests.vb [hunk 10]
```diff
@@ -410,10 +444,13 @@ End Module]]>.Value
             Dim expectedOperationTree = <![CDATA[
 IAnonymousObjectCreationExpression (OperationKind.AnonymousObjectCreationExpression, Type: <anonymous type: Key a As ?>, IsInvalid) (Syntax: 'New With {K ... essOf c1.S}')
   Initializers(1):
-      ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: ?, IsInvalid) (Syntax: 'Key .a = AddressOf c1.S')
-        Left: IPropertyReferenceExpression: ReadOnly Property <anonymous type: Key a As ?>.a As ? (Static) (OperationKind.PropertyReferenceExpression, Type: ?) (Syntax: 'a')
-            Instance Receiver: null
-        Right: IInvalidExpression (OperationKind.InvalidExpression, Type: ?, IsInvalid) (Syntax: 'AddressOf c1.S')
+      ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: ?, IsInvalid, IsImplicit) (Syntax: 'Key .a = AddressOf c1.S')
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests.vb]: >ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: ?, IsInvalid, IsImplicit) (Syntax: 'Key .a = AddressOf c1.S') [](start = 6, length = 138)

This node probably shouldn't be marked as implicit.

**Comment by AlekseyTs** [comment]
>       IArgument (ArgumentKind.DefaultValue, Matching Parameter: a) (OperationKind.Argument, IsImplicit) (Syntax: 'M2')

This node probably shouldn't be marked as implicit, there is an OmittedArgumentSyntax for it.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArgument.vb:199 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>       IArgument (ArgumentKind.Explicit, Matching Parameter: a) (OperationKind.Argument, IsImplicit) (Syntax: 'x')

This node probably shouldn't be marked as implicit.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArgument.vb:467 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>       IArgument (ArgumentKind.Explicit, Matching Parameter: a) (OperationKind.Argument, IsImplicit) (Syntax: 'x')

This node probably shouldn't be marked as implicit, there is an argument syntax for it.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArgument.vb:887 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>           IArgument (ArgumentKind.DefaultValue, Matching Parameter: b) (OperationKind.Argument, IsImplicit) (Syntax: 'M2')

This node probably shouldn't be marked as implicit, there is an OmittedArgumentSyntax for it.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IArgument.vb:1223 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                       IOperation:  (OperationKind.None, IsImplicit) (Syntax: 'x')

There is something wrong with this baseline, OperationKind.None is not expected and the conversion on top of it should probably be marked as IsImplicit.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.vb:1211 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>       ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: System.Int32) (Syntax: 'x /= y')

I would expect ICompoundAssignmentExpression node here.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBinaryOperatorExpression.vb:1199 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>         IPlaceholderExpression (OperationKind.None, IsImplicit) (Syntax: 'o?.ToString()')

I believe we have a special node to represent a reference to IConditionalAccessExpression.Expression within IConditionalAccessExpression.WhenNotNull.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConditionalAccessExpression.vb:31 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>               IReturnStatement (OperationKind.ReturnStatement) (Syntax: 'End Sub')

It looks like we have two statements associated with 'End Sub' syntax: ILabeledStatement and IReturnStatement. I think at least one of them should be marked as implicit (probably ILabeledStatement). Or both? 

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:820 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>       IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Func(Of System.Int64), IsInvalid) (Syntax: 'AddressOf M2')

Should this node be marked as IsImplicit?

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:1205 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>       IConversionExpression (Implicit, TryCast: False, Unchecked) (OperationKind.ConversionExpression, Type: System.Array, IsInvalid) (Syntax: 'New Object')

I believe this node should be marked as IsImplicit.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IConversionExpression.vb:1356 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     Partial Public Class IOperationTests

We should have tests for AddHandler/RemoveHandler when the second operand is a delegate instance or a lambda.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IEventAssignmentExpression.vb:9 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: System.Void) (Syntax: 'Mid(str, st ... ngth) = str')

This is not a simple assignment, this is a "Mid" assignment statement, it should have special representation.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb:684 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     IOperation:  (OperationKind.None, IsImplicit) (Syntax: 'x')

OperationKind.None is unexpected. Is there an issue that tracks this? 

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IParameterReferenceExpression.vb:871 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                   null

This looks wrong, the receiver shouldn't be null. I would expect IInstanceReferenceExpression node or something similar.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/AnonymousTypesTests.vb:33 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>       ISimpleAssignmentExpression (OperationKind.SimpleAssignmentExpression, Type: ?, IsInvalid, IsImplicit) (Syntax: '.a = SubName()')

IsImplicit looks unexpected.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/AnonymousTypesTests.vb:481 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                       IBlockStatement (1 statements) (OperationKind.BlockStatement) (Syntax: 'From s In q')

This node should be marked as IsImplicit.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb:82 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                         IReturnStatement (OperationKind.ReturnStatement) (Syntax: 'From s In q')

This node should be marked as IsImplicit.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb:83 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                                   IPropertyReferenceExpression: ReadOnly Property <anonymous type: Key s1 As System.Int32, Key s2 As System.Int32>.s1 As System.Int32 (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: 's2 + s1')

I think this node should be marked as implicit. 

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb:5799 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                                       IParameterReferenceExpression: $VB$It (OperationKind.ParameterReferenceExpression, Type: <anonymous type: Key s1 As System.Int32, Key s2 As System.Int32>) (Syntax: 's2 + s1')

I think this node should be marked as implicit.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb:5801 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                                           IParameterReferenceExpression: $VB$It (OperationKind.ParameterReferenceExpression, Type: <anonymous type: Key s1 As System.Int32, Key s2 As System.Int32>) (Syntax: 's2 + s1')

I think this node should be marked as implicit.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb:5809 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                                           IParameterReferenceExpression: $VB$It (OperationKind.ParameterReferenceExpression, Type: <anonymous type: Key s1 As System.Int32, Key s2 As System.Int32>) (Syntax: 's2 + s1')

I think this node should be marked as implicit.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb:5813 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                                       IPropertyReferenceExpression: ReadOnly Property <anonymous type: Key s1 As System.Int32, Key s2 As System.Int32>.s2 As System.Int32 (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: 's2 + s1')

I think the syntax for this node should be 's2'.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb:5807 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>                                       IPropertyReferenceExpression: ReadOnly Property <anonymous type: Key s1 As System.Int32, Key s2 As System.Int32>.s1 As System.Int32 (OperationKind.PropertyReferenceExpression, Type: System.Int32) (Syntax: 's2 + s1')

I think the syntax for this node should be 's1'.

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb:5811 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
>     Public Class QueryExpressions

It looks like we are lacking tests that verify nodes returned for constituent parts of a query expression. This means we are not verifying that descendants are associated with the right syntax nodes. 

---
Refers to: src/Compilers/VisualBasic/Test/Semantic/Semantics/QueryExpressions.vb:10 in 57fc445. [](commit_id = 57fc445c5005e6244e2f5ff96c03788a0252722a, deletion_comment = False)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 9).

**Comment by jinujoseph** [comment]
@AlekseyTs , As discussed ...going to merge this change and file issues to track all the discrepancies you found 
cc/ @jaredpar , @dotnet/analyzer-ioperation

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 11]
```diff
@@ -89,6 +89,13 @@ private void LogCommonPropertiesAndNewLine(IOperation operation)
                 LogString(", IsInvalid");
             }
 
+            // IsImplicit
+            if (operation.IsImplicit)
+            {
+                LogString(", IsImplicit");
+            }
+
```
Comment (heejaechang) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: extra line

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs [hunk 12]
```diff
@@ -233,8 +240,10 @@ public override void Visit(IOperation operation)
         {
             if (operation == null)
             {
```
Comment (heejaechang) [inline src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: can we log parent if it is first operation in the tree? (such as indent == 0)

File: src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs (hunk 11 above)
Comment (heejaechang) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: where is language logged?

Comment (jinujoseph) [reply src/Test/Utilities/Portable/Compilation/OperationTreeVerifier.cs]: its an assert now

**Pull request merged**